flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java

```diff
@@ -0,0 +1,210 @@
+            throws IOException {
+        this.fileSystem = fileSystem;
+        this.basePath = basePath;
+        this.deleteOnCommit = deleteOnCommit;
+
+        this.fileSystem.mkdirs(this.basePath);
+    }
+
+    public static FileSystemJobResultStore fromConfiguration(Configuration config)
+            throws IOException {
+        Preconditions.checkNotNull(config);
+        final String jrsStoragePath = config.get(JobResultStoreOptions.STORAGE_PATH);
+        final Path basePath;
+
+        if (isNullOrWhitespaceOnly(jrsStoragePath)) {
+            final String haStoragePath = config.get(HighAvailabilityOptions.HA_STORAGE_PATH);
+            final String haClusterId = config.get(HighAvailabilityOptions.HA_CLUSTER_ID);
+            basePath = new Path(haStoragePath + "/job-result-store/" + haClusterId);
+        } else {
+            basePath = new Path(jrsStoragePath);
+        }
+
+        boolean deleteOnCommit = config.get(JobResultStoreOptions.DELETE_ON_COMMIT);
+
+        return new FileSystemJobResultStore(basePath.getFileSystem(), basePath, deleteOnCommit);
+    }
+
+    /**
+     * Given a job ID, construct the path for a dirty entry corresponding to it in the job result
+     * store.
+     *
+     * @param jobId The job ID to construct a dirty entry path from.
+     * @return A path for a dirty entry for the given the Job ID.
+     */
+    private Path constructDirtyPath(JobID jobId) {
+        return new Path(this.basePath.getPath() + "/" + jobId.toString() + DIRTY_SUFFIX + ".json");
+    }
+
+    /**
+     * Given a job ID, construct the path for a clean entry corresponding to it in the job result
+     * store.
+     *
+     * @param jobId The job ID to construct a clean entry path from.
+     * @return A path for a clean entry for the given the Job ID.
+     */
+    private Path constructCleanPath(JobID jobId) {
+        return new Path(this.basePath.getPath() + "/" + jobId.toString() + ".json");
+    }
+
+    @Override
+    public void createDirtyResultInternal(JobResultEntry jobResultEntry) throws IOException {
+        final Path path = constructDirtyPath(jobResultEntry.getJobId());
+        OutputStream os = fileSystem.create(path, FileSystem.WriteMode.NO_OVERWRITE);
+        mapper.writeValue(os, new JsonJobResultEntry(jobResultEntry));
+    }
+
+    @Override
+    public void markResultAsCleanInternal(JobID jobId) throws IOException, NoSuchElementException {
+        Path dirtyPath = constructDirtyPath(jobId);
+        if (fileSystem.exists(dirtyPath)) {
+            if (deleteOnCommit) {
+                fileSystem.delete(dirtyPath, false);
+            } else {
+                fileSystem.rename(dirtyPath, constructCleanPath(jobId));
+            }
+        } else {
+            throw new NoSuchElementException(
+                    String.format(
+                            "Could not mark job %s as clean as it is not present in the job result store.",
+                            jobId));
+        }
```

Comment:
```suggestion
        if (!fileSystem.exists(dirtyPath)) {
            throw new NoSuchElementException(
                    String.format(
                            "Could not mark job %s as clean as it is not present in the job result store.",
                            jobId));
        }
        
        if (deleteOnCommit) {
            fileSystem.delete(dirtyPath, false);
        } else {
            fileSystem.rename(dirtyPath, constructCleanPath(jobId));
        }
```
nit: You could remove the nested if by inverting the if condition which might improve readability. But again, like most of the other comments - it's more like a cosmetic change and up to your taste.